dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/reader/ExternalReader.java

```diff
@@ -0,0 +1,150 @@
+ * this program. If not, see <http://www.gnu.org/licenses/>.                                                     
+ *                                                                                                               
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY                           
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES                          
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT                           
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,                                
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED                          
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;                               
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER                              
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN                         
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.icgc.dcc.imports.gene.reader;
+
+import static org.icgc.dcc.common.json.Jackson.DEFAULT;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.icgc.dcc.imports.gene.model.GeneMapping;
+import org.icgc.dcc.imports.gene.model.TranslationMapping;
+import org.icgc.dcc.imports.gene.model.XrefMapping;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableMap;
+
+import lombok.NonNull;
+import lombok.val;
+
+/**
+ * Responsible for creating a map of gene to external db ids
+ */
+public final class ExternalReader extends TsvReader {
+
+  /**
+   * Dependencies
+   */
+  private final XrefMapping xrefMapping;
+  private final Map<String, String> getIdMap;
+  private final Map<String, String> translationToGene;
+
+  public ExternalReader(String uri,
+      @NonNull XrefMapping xrefMapping,
+      @NonNull GeneMapping geneMapping,
+      @NonNull TranslationMapping translationMapping) {
+    super(uri);
+    this.xrefMapping = xrefMapping;
+    this.getIdMap = geneMapping.getGeneIdMap();
+    this.translationToGene = translationMapping.getTranslationToGene();
+
+  }
+
+  public Map<String, ObjectNode> read() {
+    val externalIds = new HashMap<String, ObjectNode>();
+
+    readRecords().forEach(record -> {
+      if (isGene(record)) {
+        handleExternalGeneIds(record, externalIds);
+      } else if (isTranslation(record)) {
+        handleUniprotIds(record, externalIds);
+      }
+    });
+
+    return ImmutableMap.<String, ObjectNode> copyOf(externalIds);
```

Comment:
`ImmutableMap.copyOf(externalIds);` should work due to return type filling in the details.